stablehlo/dialect/Base.cpp

```diff
@@ -176,5 +176,151 @@ Attribute boundsToEncoding(Attribute prototype, ArrayRef<int64_t> bounds) {
+  if (isRightStaticBound)
+    return std::make_pair(ShapedType::kDynamicSize, rightBound);
+  return std::make_pair(ShapedType::kDynamicSize, ShapedType::kDynamicSize);
+}
+
+LogicalResult inferMostSpecificType(
+    Optional<Location> location, ValueTypeRange<ValueRange> inputTypes,
+    SmallVectorImpl<Type>& inferredReturnTypes,
+    const SmallVector<int64_t>& concatenateDims) {
+  SmallVector<RankedTensorType> rankedTypes;
+  for (auto inputType : inputTypes)
+    if (auto rankedType = inputType.dyn_cast<RankedTensorType>())
+      rankedTypes.push_back(rankedType);
+  if (rankedTypes.empty()) {
+    inferredReturnTypes.push_back(inputTypes[0]);
+    return success();
+  }
+
+  auto rank = rankedTypes[0].getRank();
+  BoundedDialectInterface* dialect = nullptr;
+  SmallVector<int64_t> inferredDimSizes(rank, ShapedType::kDynamicSize);
+  SmallVector<int64_t> inferredBounds(rank, ShapedType::kDynamicSize);
+
+  for (const auto& it : llvm::enumerate(rankedTypes)) {
+    RankedTensorType rankedType = it.value();
+    ArrayRef<int64_t> bounds;
+    if (auto boundedAttr =
+            rankedType.getEncoding().dyn_cast_or_null<BoundedAttrInterface>()) {
+      dialect = cast<BoundedDialectInterface>(&boundedAttr.getDialect());
+      bounds = boundedAttr.getBounds();
+    } else if (rankedType.getEncoding()) {
+      // TODO(zhouxin) infer sparsity encoding after b/238903065 is fixed.
+      inferredReturnTypes.push_back(inputTypes[0]);
+      return success();
+    }
+
+    // Init the inferredDimSizes[dim] / inferredBounds[dim] with rankedTypes[0]
+    if (it.index() == 0) {
+      for (int dim = 0; dim < rank; ++dim) {
+        inferredDimSizes[dim] = rankedType.getDimSize(dim);
+        inferredBounds[dim] =
+            bounds.empty() ? ShapedType::kDynamicSize : bounds[dim];
+      }
+      continue;
+    }
+
+    // Infer each dim for current rankedType (from inputTypes[1:end])
+    for (int dim = 0; dim < rank; ++dim) {
+      std::pair<int64_t, int64_t> inferredDimAndBound;
+
+      int64_t leftSize = inferredDimSizes[dim];
+      int64_t rightSize = rankedType.getShape()[dim];
+      int64_t leftBound = inferredBounds[dim];
+      int64_t rightBound =
+          bounds.empty() ? ShapedType::kDynamicSize : bounds[dim];
+      if (std::find(concatenateDims.begin(), concatenateDims.end(), dim) !=
+          concatenateDims.end()) {
+        inferredDimAndBound = inferConcatenatedDimAndBound(
+            leftSize, rightSize, leftBound, rightBound);
+      } else {
+        auto inferredDimAndBoundOrErr = inferMergedDimAndBound(
+            location, dim, leftSize, rightSize, leftBound, rightBound);
+        if (failed(inferredDimAndBoundOrErr)) return failure();
+        inferredDimAndBound = *inferredDimAndBoundOrErr;
+      }
+      inferredDimSizes[dim] = inferredDimAndBound.first;
+      inferredBounds[dim] = inferredDimAndBound.second;
+    }
+  }
+
+  Attribute encoding = nullptr;
+  if (llvm::any_of(inferredBounds,
```

Comment:
we could have a helper function to get encoding attribute from bounds? it will be a tiny helper but will improve the code readability.

Reply:
New commit is much refined with our existing `encodingToBounds() / boundsToEncoding()` utils.